{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import yaml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/Users/Jacob/Desktop/test/gdcdictionary-2.2.0/gdcdictionary/schemas/follow_up.yaml') as file:\n",
    "    test = yaml.load(file,Loader = yaml.FullLoader)\n",
    "with open('/Users/Jacob/Desktop/test/gdcdictionary-2.1.0/gdcdictionary/schemas/follow_up.yaml') as file:\n",
    "    oldtest = yaml.load(file,Loader = yaml.FullLoader)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('diagnosis.yaml') as file:\n",
    "    test = yaml.load(file,Loader = yaml.FullLoader)\n",
    "with open('olddiagnosis.yaml') as file:\n",
    "    oldtest = yaml.load(file,Loader = yaml.FullLoader)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'exclusive': False,\n",
       " 'required': True,\n",
       " 'subgroup': [{'name': 'aggregated_somatic_mutations',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'aggregated_somatic_mutation',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'aligned_reads_indexes',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'aligned_reads_index',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'aligned_reads_files',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'aligned_reads',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'aliquots',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'aliquot',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'analysis_metadata_files',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'analysis_metadata',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'analytes',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'analyte',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'annotated_somatic_mutations',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'annotated_somatic_mutation',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'archives',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'archive',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'biospecimen_supplements',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'biospecimen_supplement',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'cases',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'case',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'centers',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'center',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'clinical_supplements',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'clinical_supplement',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'copy_number_estimates',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'copy_number_estimate',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'copy_number_segments',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'copy_number_segment',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'demographics',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'demographic',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'diagnoses',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'diagnosis',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'experiment_metadata_files',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'experiment_metadata',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'exposures',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'exposure',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'family_histories',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'family_history',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'files',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'file',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'filtered_copy_number_segments',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'filtered_copy_number_segment',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'follow_ups',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'follow_up',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'gene_expressions',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'gene_expression',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'masked_somatic_mutations',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'masked_somatic_mutation',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'methylation_beta_values',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'methylation_beta_value',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'mirna_expressions',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'mirna_expression',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'pathology_reports',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'pathology_report',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'portions',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'portion',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'protein_expressions',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'protein_expression',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'raw_methylation_arrays',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'raw_methylation_array',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'read_group_qcs',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'read_group_qc',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'read_groups',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'read_group',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'run_metadata_files',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'run_metadata',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'samples',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'sample',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'secondary_expression_analyses',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'secondary_expression_analysis',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'simple_germline_variations',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'simple_germline_variation',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'simple_somatic_mutations',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'simple_somatic_mutation',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'slide_images',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'slide_image',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'slides',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'slide',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'somatic_mutation_indexes',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'somatic_mutation_index',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'structural_variations',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'structural_variation',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'submitted_aligned_reads_files',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'submitted_aligned_reads',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'submitted_genomic_profiles',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'submitted_genomic_profile',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'submitted_genotyping_arrays',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'submitted_genotyping_array',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'submitted_methylation_beta_values',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'submitted_methylation_beta_value',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'submitted_tangent_copy_numbers',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'submitted_tangent_copy_number',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'submitted_unaligned_reads_files',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'submitted_unaligned_reads',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'tissue_source_sites',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'tissue_source_site',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False},\n",
       "  {'name': 'treatments',\n",
       "   'backref': 'annotations',\n",
       "   'label': 'annotates',\n",
       "   'target_type': 'treatment',\n",
       "   'multiplicity': 'many_to_many',\n",
       "   'required': False}]}"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test['links'][0][]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test.keys()==oldtest.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "def findDifferences(dict1,dict2,it = 0,*args):\n",
    "    done = False\n",
    "    if type(dict1)==dict and type(dict2)==dict:\n",
    "        for key in dict2.keys():\n",
    "            if key == 'enumDef':\n",
    "                continue\n",
    "            elif key not in dict1.keys(): # New Keys\n",
    "                for arg in args:\n",
    "                    if done==False:\n",
    "                        print('\\t'*(it) + '* Changes made to ' + arg)\n",
    "                        done = True\n",
    "                print('\\t'*(it+1) + '* New property:',key)\n",
    "                \n",
    "                \n",
    "            elif type(dict2[key])==list: # Executes if item in dictionary is a list\n",
    "                if dict2[key]!=dict1[key]: # Only executes next section of code if two lists are unequal\n",
    "                    for arg in args:\n",
    "                        print('\\t'*(it) + '* Changes made to ' + arg)\n",
    "                    if key != 'enum':\n",
    "                        print('\\t'*(it+1) + '* Changes made to ' + key)\n",
    "                    for item in dict2[key]:\n",
    "                        if type(item)==dict: # Iterates through dictionaries nested in lists\n",
    "                            findDifferences(dict1[key],dict2[key],it + 1,str(key))\n",
    "                        elif item not in dict1[key]: # Checks for items in the new test \n",
    "                            if key !='enum':\n",
    "                                print('\\t'*(it+2) + '* New Item:',item)\n",
    "                            else:\n",
    "                                print('\\t'*(it+1) + '* New permissable value: ' + item)\n",
    "                    for item in dict1[key]:\n",
    "                        if item not in dict2[key]:\n",
    "                            print('\\t'*(it+2) + '* Dropped Item:',item) \n",
    "                            \n",
    "                            \n",
    "        for key in dict2.keys():# This is put down here for formatting issues\n",
    "            if key == 'enumDef':\n",
    "                continue\n",
    "            elif key not in dict1.keys(): # New Keys\n",
    "                continue\n",
    "                \n",
    "            elif type(dict2[key])==dict and dict2[key]!=dict1[key]: # Recursion through nested Dictionaries\n",
    "                findDifferences(dict1[key],dict2[key],it + 1,str(key)) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "def findDifferences(dict1,dict2,it = 0,*args):\n",
    "    done = False\n",
    "    if type(dict1)==dict and type(dict2)==dict:\n",
    "        for key in dict2.keys():\n",
    "            if key == 'enumDef':\n",
    "                continue\n",
    "                \n",
    "            elif key == 'links':\n",
    "                try:\n",
    "                    for item in dict2[key][0]['subgroup']:\n",
    "                        if type(item)==dict and item not in dict1[key][0]['subgroup']:\n",
    "                            for item1 in dict1[key][0]['subgroup']:\n",
    "                                if item['name']==item1['name']:\n",
    "                                        findDifferences(item,item1)\n",
    "                            else:\n",
    "                                print('\\t'*(it+1) + '* Changes made to ' + key)\n",
    "                                print('\\t'*(it+2)+'* '+item['name']+' added to subgroup')\n",
    "                except Exception:\n",
    "                    pass\n",
    "            \n",
    "            elif key not in dict1.keys(): # New Keys\n",
    "                for arg in args:\n",
    "                    if done==False:\n",
    "                        print('\\t'*(it) + '* Changes made to ' + arg)\n",
    "                        done = True\n",
    "                print('\\t'*(it+1) + '* New property:',key)\n",
    "                \n",
    "                \n",
    "            elif type(dict2[key])==list: # Executes if item in dictionary is a list\n",
    "                if dict2[key]!=dict1[key]: # Only executes next section of code if two lists are unequal\n",
    "                    for arg in args:\n",
    "                        print('\\t'*(it) + '* Changes made to ' + arg)\n",
    "                    if key != 'enum':\n",
    "                        print('\\t'*(it+1) + '* Changes made to ' + key)\n",
    "                    for item in dict2[key]:\n",
    "                        if type(item)==dict and item not in dict1[key]: # Iterates through dictionaries nested in lists\n",
    "                            findDifferences(dict1[key],dict2[key],it + 1,key)\n",
    "                        elif item not in dict1[key]: # Checks for items in the new test \n",
    "                            if key !='enum':\n",
    "                                print('\\t'*(it+2) + '* New Item:',item)\n",
    "                            else:\n",
    "                                print('\\t'*(it+1) + '* New permissable value: ' + item)\n",
    "                    for item in dict1[key]:\n",
    "                        if item not in dict2[key]:\n",
    "                            print('\\t'*(it+2) + '* Dropped Item:',item) \n",
    "                            \n",
    "                            \n",
    "        for key in dict2.keys():# This is put down here for formatting issues\n",
    "            if key == 'enumDef':\n",
    "                continue\n",
    "            elif key not in dict1.keys(): # New Keys\n",
    "                continue\n",
    "                \n",
    "            elif type(dict2[key])==dict and dict2[key]!=dict1[key]: # Recursion through nested Dictionaries\n",
    "                findDifferences(dict1[key],dict2[key],it + 1,str(key)) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-137-bd96a18f7933>, line 55)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-137-bd96a18f7933>\"\u001b[0;36m, line \u001b[0;32m55\u001b[0m\n\u001b[0;31m    if !done:\u001b[0m\n\u001b[0m       ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "def findDifferences(dict1,dict2,it = 0,*args):\n",
    "    done = False\n",
    "    if type(dict1)==dict and type(dict2)==dict:\n",
    "        for key in dict2.keys():\n",
    "            if key == 'enumDef':\n",
    "                continue\n",
    "                \n",
    "            elif key == 'links':\n",
    "                try:\n",
    "                    for item in dict2[key][0]['subgroup']:\n",
    "                        if type(item)==dict and item not in dict1[key][0]['subgroup']:\n",
    "                            for item1 in dict1[key][0]['subgroup']:\n",
    "                                if item['name']==item1['name']:\n",
    "                                        findDifferences(item,item1)\n",
    "                            else:\n",
    "                                file.write('\\t'*(it+1) + '* Changes made to ' + key + '\\n')\n",
    "                                file.write('\\t'*(it+2)+'* '+item['name']+' added to subgroup' + '\\n')\n",
    "                except Exception:\n",
    "                    pass\n",
    "                \n",
    "            elif key not in dict1.keys(): # New Keys\n",
    "                for arg in args:\n",
    "                    if done==False:\n",
    "                        file.write('\\t'*(it) + '* Changes made to ' + arg + '\\n')\n",
    "                        done = True\n",
    "                file.write('\\t'*(it+1) + '* New property: ' + key + '\\n')\n",
    "                \n",
    "                \n",
    "            elif type(dict2[key])==list: # Executes if item in dictionary is a list\n",
    "                if dict2[key]!=dict1[key]: # Only executes next section of code if two lists are unequal\n",
    "                    for arg in args:\n",
    "                        file.write('\\t'*(it) + '* Changes made to ' + arg + '\\n')\n",
    "                    if key != 'enum':\n",
    "                        file.write('\\t'*(it+1) + '* Changes made to ' + key + '\\n')\n",
    "                    for item in dict2[key]:\n",
    "                        if type(item)==dict: # Iterates through dictionaries nested in lists\n",
    "                            findDifferences(dict1[key],dict2[key],it + 1,str(key))\n",
    "                        elif item not in dict1[key]: # Checks for items in the new test \n",
    "                            if key !='enum':\n",
    "                                file.write('\\t'*(it+2) + '* New Item: ' + item + '\\n')\n",
    "                            else:\n",
    "                                file.write('\\t'*(it+1) + '* New permissable value: ' + item + '\\n')\n",
    "                    for item in dict1[key]:\n",
    "                        if item not in dict2[key]:\n",
    "                            file.write('\\t'*(it+2) + '* Dropped Item: ' + item + '\\n') \n",
    "                            \n",
    "                            \n",
    "        for key in dict2.keys():# This is put down here for formatting issues\n",
    "            if key == 'enumDef':\n",
    "                continue\n",
    "            elif key not in dict1.keys(): # New Keys\n",
    "                continue\n",
    "                \n",
    "            elif type(dict2[key])==dict and dict2[key]!=dict1[key]: # Recursion through nested Dictionaries\n",
    "                if done == False:\n",
    "                    for arg in args:\n",
    "                        file.write('\\t'*it + '* Changes made to ' + arg + '\\n')\n",
    "                findDifferences(dict1[key],dict2[key],it + 1,str(key)) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "def isDifferent(dict1,dict2, truth = False):\n",
    "    if truth:\n",
    "        return True\n",
    "    if type(dict1)==dict and type(dict2)==dict:\n",
    "        if truth:\n",
    "            return True\n",
    "        for key in dict2.keys():\n",
    "            if key == 'enumDef':\n",
    "                continue\n",
    "            elif key not in dict1.keys(): # New Keys\n",
    "                return True\n",
    "            elif key!='properties' and dict1[key]!=dict2[key]:\n",
    "                return True\n",
    "                \n",
    "                \n",
    "            if type(dict2[key])==list: # Executes if item in dictionary is a list\n",
    "                if dict2[key]!=dict1[key]: # Only executes next section of code if two lists are unequal\n",
    "                    return True\n",
    "                            \n",
    "                            \n",
    "        for key in dict2.keys():# This is put down here for formatting issues\n",
    "            if truth:\n",
    "                return truth\n",
    "            if key == 'enumDef':\n",
    "                continue\n",
    "            elif key not in dict1.keys(): # New Keys\n",
    "                return True\n",
    "                \n",
    "            elif type(dict2[key])==dict and dict2[key]!=dict1[key]: # Recursion through nested Dictionaries\n",
    "                truth = isDifferent(dict1[key],dict2[key],truth)\n",
    "    else:\n",
    "        return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t* Changes made to properties\n",
      "\t\t* New property: hormonal_contraceptive_type\n",
      "\t\t* New property: hormone_replacement_therapy_type\n",
      "\t\t* New property: procedures_performed\n",
      "\t\t* Changes made to aids_risk_factors\n",
      "\t\t\t* New permissable value: Cryptosporidiosis, Chronic Intestinal\n",
      "\t\t* Changes made to comorbidity\n",
      "\t\t\t* New permissable value: Ataxia-telangiectasia\n",
      "\t\t\t* New permissable value: Autoimmune Lymphoproliferative Syndrome (ALPS)\n",
      "\t\t\t* New permissable value: Bacteroides fragilis\n",
      "\t\t\t* New permissable value: Chlamydia\n",
      "\t\t\t* New permissable value: Chronic Systemic Steroid Use\n",
      "\t\t\t* New permissable value: Cryptococcal Meningitis\n",
      "\t\t\t* New permissable value: Cytomegalovirus (CMV)\n",
      "\t\t\t* New permissable value: Lymphocytic Meningitis\n",
      "\t\t\t* New permissable value: Malaria\n",
      "\t\t\t* New permissable value: Metabolic Syndrome\n",
      "\t\t\t* New permissable value: Mycobacterium avium Complex\n",
      "\t\t\t* New permissable value: Pneumocystis Pneumonia\n",
      "\t\t\t* New permissable value: Polycystic Ovarian Syndrome (PCOS)\n",
      "\t\t\t* New permissable value: Shingles\n",
      "\t\t\t* New permissable value: Sjogren's Syndrome\n",
      "\t\t\t* New permissable value: Staphylococcus aureus\n",
      "\t\t\t* New permissable value: Syphilis\n",
      "\t\t\t* New permissable value: Treponema pallidum\n",
      "\t\t* Changes made to evidence_of_recurrence_type\n",
      "\t\t\t* New permissable value: Physical Examination\n",
      "\t\t* Changes made to risk_factor\n",
      "\t\t\t* New permissable value: Ataxia-telangiectasia\n",
      "\t\t\t* New permissable value: Autoimmune Lymphoproliferative Syndrome (ALPS)\n",
      "\t\t\t* New permissable value: Bacteroides fragilis\n",
      "\t\t\t* New permissable value: BAP1 Tumor Predisposition Syndrome\n",
      "\t\t\t* New permissable value: Birt-Hogg-Dube Syndrome\n",
      "\t\t\t* New permissable value: Chlamydia\n",
      "\t\t\t* New permissable value: Chronic Systemic Steroid Use\n",
      "\t\t\t* New permissable value: Cowden Syndrome\n",
      "\t\t\t* New permissable value: Cryptococcal Meningitis\n",
      "\t\t\t* New permissable value: Cytomegalovirus (CMV)\n",
      "\t\t\t* New permissable value: Hereditary Breast Cancer\n",
      "\t\t\t* New permissable value: Hereditary Kidney Oncocytoma\n",
      "\t\t\t* New permissable value: Hereditary Leiomyomatosis and Renal Cell Carcinoma\n",
      "\t\t\t* New permissable value: Hereditary Ovarian Cancer\n",
      "\t\t\t* New permissable value: Hereditary Papillary Renal Cell Carcinoma\n",
      "\t\t\t* New permissable value: Hereditary Renal Cell Carcinoma\n",
      "\t\t\t* New permissable value: Lymphocytic Meningitis\n",
      "\t\t\t* New permissable value: Malaria\n",
      "\t\t\t* New permissable value: Metabolic Syndrome\n",
      "\t\t\t* New permissable value: Mycobacterium avium Complex\n",
      "\t\t\t* New permissable value: Pneumocystis Pneumonia\n",
      "\t\t\t* New permissable value: Polycystic Ovarian Syndrome (PCOS)\n",
      "\t\t\t* New permissable value: Shingles\n",
      "\t\t\t* New permissable value: Sjogren's Syndrome\n",
      "\t\t\t* New permissable value: Staphylococcus aureus\n",
      "\t\t\t* New permissable value: Succinate Dehydrogenase-Deficient Renal Cell Carcinoma\n",
      "\t\t\t* New permissable value: Syphilis\n",
      "\t\t\t* New permissable value: Treponema pallidum\n",
      "\t\t\t* New permissable value: Tuberculosis\n",
      "\t\t\t* New permissable value: Tuberous Sclerosis\n",
      "\t\t\t* New permissable value: Von Hippel-Lindau Syndrome\n"
     ]
    }
   ],
   "source": [
    "findDifferences(oldtest,test)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Hello\n",
    "    * Changes made to required\n",
    "\t\t* Dropped Item: days_to_last_follow_up\n",
    "\t\t* Dropped Item: tumor_grade\n",
    "\t\t* Dropped Item: progression_or_recurrence\n",
    "\t\t* Dropped Item: days_to_recurrence\n",
    "\t\t* Dropped Item: days_to_last_known_disease_status\n",
    "\t\t* Dropped Item: last_known_disease_status\n",
    "\t* Changes made to deprecated\n",
    "\t\t* New Item: micropapillary_features\n",
    "\t\t* New Item: mitotic_count\n",
    "\t\t* New Item: non_nodal_regional_disease\n",
    "\t\t* New Item: non_nodal_tumor_deposits\n",
    "\t\t* New Item: papillary_renal_cell_type\n",
    "\t* Changes made to properties\n",
    "\t\t* New property: eln_risk_classification\n",
    "\t\t* New property: satellite_nodule_present\n",
    "\t\t* New property: sites_of_involvement\n",
    "\t\t* New property: who_cns_grade\n",
    "\t\t* New property: who_nte_grade\n",
    "\t\t* Changes made to ajcc_pathologic_stage\n",
    "\t\t\t* New permissable value: Stage IA3\n",
    "\t\t* Changes made to classification_of_tumor\n",
    "\t\t\t* New permissable value: Progression\n",
    "\t\t* Changes made to metastasis_at_diagnosis_site\n",
    "\t\t\t* New permissable value: Esophagus\n",
    "\t\t* Changes made to morphology\n",
    "\t\t\t* New permissable value: 8040/3\n",
    "\t\t\t* New permissable value: 8046/6\n",
    "\t\t\t* New permissable value: 8240/6\n",
    "\t\t\t* New permissable value: 8471/1\n",
    "\t\t\t* New permissable value: 8500/6\n",
    "\t\t\t* New permissable value: 8720/6\n",
    "\t\t\t* New permissable value: 8801/6\n",
    "\t\t\t* New permissable value: 8804/6\n",
    "\t\t\t* New permissable value: 8920/6\n",
    "\t\t\t* New permissable value: 8950/6\n",
    "\t\t\t* New permissable value: 9180/6\n",
    "\t\t\t* New permissable value: 9440/6\n",
    "\t\t* Changes made to site_of_resection_or_biopsy\n",
    "\t\t\t* New permissable value: Pancreatic neck\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('test.md','w') as file:\n",
    "    file.write('* Hi\\n\\t* Ho')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "isDifferent(oldtest,test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isDifferent(oldtest,test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "def isDifferent(dict1,dict2,it = 0,truth = False,*args):\n",
    "    done = False\n",
    "    if truth:\n",
    "        return True\n",
    "    if type(dict1)==dict and type(dict2)==dict:\n",
    "        for key in dict2.keys():\n",
    "            if truth:\n",
    "                return True\n",
    "            if key == 'enumDef':\n",
    "                continue\n",
    "                \n",
    "            elif key == 'links':\n",
    "                try:\n",
    "                    for item in dict2[key][0]['subgroup']:\n",
    "                        if type(item)==dict and item not in dict1[key][0]['subgroup']:\n",
    "                            return True\n",
    "                except Exception:\n",
    "                    pass\n",
    "            \n",
    "            if key not in dict1.keys(): # New Keys\n",
    "                return True\n",
    "                \n",
    "            if type(dict2[key])==list: # Executes if item in dictionary is a list\n",
    "                if dict2[key]!=dict1[key]: # Only executes next section of code if two lists are unequal\n",
    "                    return True\n",
    "                            \n",
    "                            \n",
    "        for key in dict2.keys():# This is put down here for formatting issues\n",
    "            if truth:\n",
    "                return True\n",
    "            if key == 'enumDef':\n",
    "                continue\n",
    "            if key not in dict1.keys(): # New Keys\n",
    "                return True\n",
    "                \n",
    "            if type(dict2[key])==dict: # Recursion through nested Dictionaries\n",
    "                truth = isDifferent(dict1[key],dict2[key],truth)\n",
    "                if truth:\n",
    "                    return True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
